compiler/pipes/final-check.cpp

```diff
@@ -1063,3 +1065,49 @@ void FinalCheckPass::check_magic_clone_method(FunctionPtr fun) {
   kphp_error(!fun->is_resumable, fmt_format("{} method has to be not resumable", fun->as_human_readable()));
   kphp_error(!fun->can_throw(), fmt_format("{} method should not throw exception", fun->as_human_readable()));
 }
+
+
+void FinalCheckPass::check_serialized_fields_hierarchy(ClassPtr klass) {
+  // Protect from multiple error reporting for class hierarchy
+  bool instance_fields_error_reported = false;
+
+  klass->members.for_each([&](ClassMemberInstanceField &f) {
+    auto the_klass = klass;
+
+    // This loop finishes unconditionally since there is NULL klass->parent_class if there is no base class.
+    while (the_klass) {
+
+      // Inheritance with serialization is allowed if
+      // * parent class has ho instance field
+      // * if there are instance fields, class should be marked with kphp-serializable
+      if (klass->is_serializable && !instance_fields_error_reported) {
+
+        bool hierarchy_condition = (the_klass->members.has_any_instance_var() && the_klass->is_serializable) || !the_klass->members.has_any_instance_var();
+        if (!hierarchy_condition) {
+          instance_fields_error_reported = true;
+        }
```

Comment:
I don't understand what is `instance_fields_error_reported` for?
We can just break the loop once `hierarchy_condition == false`

Reply:
I have eliminated this check at all.